system-addon/lib/TopSitesFeed.jsm

```diff
@@ -180,14 +191,36 @@ this.TopSitesFeed = class TopSitesFeed {
     this._requestRichIcon(link.url);
 
     // Also request a screenshot if we don't have one yet
-    if (!link.screenshot) {
-      const {url} = link;
-      await Screenshots.maybeCacheScreenshot(link, url, "screenshot",
-        screenshot => this.store.dispatch(ac.BroadcastToContent({
-          data: {screenshot, url},
-          type: at.SCREENSHOT_UPDATED
-        })));
+    await this._fetchScreenshot(link, link.url);
+  }
+
+  /**
+   * Fetch, cache and broadcast a screenshot for a specific topsite.
+   * @param link cached topsite object
+   * @param url where to fetch the image from
+   */
+  async _fetchScreenshot(link, url) {
+    if (link.screenshot) {
+      return;
     }
+    await Screenshots.maybeCacheScreenshot(link, url, "screenshot",
+      screenshot => this.store.dispatch(ac.BroadcastToContent({
+        data: {screenshot, url: link.url},
+        type: at.SCREENSHOT_UPDATED
+      })));
+  }
+
+  /**
+   * Dispatch screenshot preview to target or notify if request failed.
+   * @param customScreenshotURL {string} The URL used to capture the screenshot
+   * @param target {string} Id of content process where to dispatch the result
+   */
+  async getScreenshotPreview({customScreenshotURL}, target) {
+    const screenshotPreview = await Screenshots.getScreenshotForURL(customScreenshotURL);
+    this.store.dispatch(ac.OnlyToOneContent({
+      data: {screenshotPreview, customScreenshotURL},
+      type: screenshotPreview ? at.SCREENSHOT_PREVIEW : at.PREVIEW_FAILED
```

Comment:
To match PREVIEW_REQUEST's {url}, this should dispatch `data: {url, preview}, type: PREVIEW_RESPONSE` where `preview` is `""` for failed.